File: sdks/ts/memory/src/ingest/directory.test.ts
// SPDX-License-Identifier: Apache-2.0

import { mkdir, symlink, writeFile } from 'node:fs/promises'
import { tmpdir } from 'node:os'
import { join } from 'node:path'
import { afterEach, beforeEach, describe, expect, it } from 'vitest'
import { mkdtemp, rm } from 'node:fs/promises'
import { enumerateFiles } from './directory.js'

describe('enumerateFiles', () => {
  let tmp: string

  beforeEach(async () => {
    tmp = await mkdtemp(join(tmpdir(), 'dir-enum-test-'))
  })

  afterEach(async () => {
    await rm(tmp, { recursive: true, force: true })
  })

  it('enumerates markdown files in a flat directory', async () => {
    await writeFile(join(tmp, 'a.md'), '# A\nContent A')
    await writeFile(join(tmp, 'b.md'), '# B\nContent B')
    await writeFile(join(tmp, 'c.md'), '# C\nContent C')

    const result = await enumerateFiles({ directory: tmp })
    expect(result.files).toHaveLength(3)
    expect(result.skipped).toHaveLength(0)
  })

  it('applies glob filter to only match *.md files', async () => {
    await writeFile(join(tmp, 'a.md'), 'markdown')
    await writeFile(join(tmp, 'b.txt'), 'text')

    const result = await enumerateFiles({ directory: tmp, glob: '*.md' })
    expect(result.files).toHaveLength(1)
    expect(result.files[0]?.path).toContain('a.md')
  })

  it('respects recursive=false and only returns top-level files', async () => {
    await writeFile(join(tmp, 'top.md'), 'top')
    await mkdir(join(tmp, 'sub'))
    await writeFile(join(tmp, 'sub', 'nested.md'), 'nested')

    const result = await enumerateFiles({ directory: tmp, recursive: false })
    expect(result.files).toHaveLength(1)
    expect(result.files[0]?.path).toContain('top.md')
  })

  it('respects maxFiles limit and reports skipped reason', async () => {
    await writeFile(join(tmp, 'a.md'), 'a')
    await writeFile(join(tmp, 'b.md'), 'b')
    await writeFile(join(tmp, 'c.md'), 'c')
    await writeFile(join(tmp, 'd.md'), 'd')
    await writeFile(join(tmp, 'e.md'), 'e')

    const result = await enumerateFiles({ directory: tmp, maxFiles: 2 })
    expect(result.files).toHaveLength(2)
    expect(result.skipped.some((s) => s.includes('max files limit'))).toBe(true)
  })

  it('excludes hidden files', async () => {
    await writeFile(join(tmp, '.hidden.md'), 'hidden')
    await writeFile(join(tmp, 'visible.md'), 'visible')

    const result = await enumerateFiles({ directory: tmp })
    expect(result.files).toHaveLength(1)
    expect(result.files[0]?.path).toContain('visible.md')
  })

  it('skips files with unknown extensions', async () => {
    await writeFile(join(tmp, 'image.png'), 'binary')
    await writeFile(join(tmp, 'doc.md'), 'markdown')

    const result = await enumerateFiles({ directory: tmp })
    expect(result.files).toHaveLength(1)
    expect(result.files[0]?.path).toContain('doc.md')
  })

  it('respects .gitignore patterns', async () => {
    await writeFile(join(tmp, '.gitignore'), 'node_modules\n*.log\n')
    await mkdir(join(tmp, 'node_modules'))
    await writeFile(join(tmp, 'node_modules', 'pkg.txt'), 'pkg')
    await writeFile(join(tmp, 'debug.log'), 'logs')
    await writeFile(join(tmp, 'readme.md'), 'readme')

    const result = await enumerateFiles({ directory: tmp })
    expect(result.files).toHaveLength(1)
    expect(result.files[0]?.path).toContain('readme.md')
  })

  it('handles non-existent directory gracefully', async () => {
    const result = await enumerateFiles({ directory: join(tmp, 'nonexistent') })
    expect(result.files).toHaveLength(0)
    expect(result.skipped.length).toBeGreaterThan(0)
  })

  it('walks subdirectories recursively by default', async () => {
    await mkdir(join(tmp, 'sub'))
    await writeFile(join(tmp, 'top.md'), 'top')
    await writeFile(join(tmp, 'sub', 'nested.md'), 'nested')

    const result = await enumerateFiles({ directory: tmp })
    expect(result.files).toHaveLength(2)
  })

  it('skips symlinks and only returns real files', async () => {
    await writeFile(join(tmp, 'real.md'), 'real content')
    const outsideDir = await mkdtemp(join(tmpdir(), 'dir-enum-outside-'))
    await writeFile(join(outsideDir, 'secret.md'), 'secret')
    try {
      await symlink(join(outsideDir, 'secret.md'), join(tmp, 'link.md'))
    } catch {
      // Symlinks may not be supported on all platforms; skip test.
      return
    }

    const result = await enumerateFiles({ directory: tmp })
    expect(result.files).toHaveLength(1)
    expect(result.files[0]?.path).toContain('real.md')
    await rm(outsideDir, { recursive: true, force: true })
  })

  it('maxFiles > 500 still works at library level (cap is in MCP tool layer)', async () => {
    await writeFile(join(tmp, 'a.md'), 'content')

    const result = await enumerateFiles({ directory: tmp, maxFiles: 501 })
    expect(result.files).toHaveLength(1)
  })
})

File: go/ingest/directory_test.go
// SPDX-License-Identifier: Apache-2.0

package ingest

import (
	"context"
	"os"
	"path/filepath"
	"testing"
)

func writeTestFile(t *testing.T, path, content string) {
	t.Helper()
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		t.Fatalf("mkdir: %v", err)
	}
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatalf("write: %v", err)
	}
}

func TestEnumerateFiles_FlatDirectory(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "a.md"), "# A")
	writeTestFile(t, filepath.Join(dir, "b.md"), "# B")
	writeTestFile(t, filepath.Join(dir, "c.md"), "# C")

	files, skipped, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 3 {
		t.Errorf("expected 3 files, got %d", len(files))
	}
	if len(skipped) != 0 {
		t.Errorf("expected 0 skipped, got %d: %v", len(skipped), skipped)
	}
}

func TestEnumerateFiles_GlobFilter(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "a.md"), "markdown")
	writeTestFile(t, filepath.Join(dir, "b.txt"), "text")

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Glob:      "*.md",
		Recursive: true,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file, got %d", len(files))
	}
}

func TestEnumerateFiles_NonRecursive(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "top.md"), "top")
	writeTestFile(t, filepath.Join(dir, "sub", "nested.md"), "nested")

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: false,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file, got %d", len(files))
	}
}

func TestEnumerateFiles_MaxFilesLimit(t *testing.T) {
	dir := t.TempDir()
	for i := 0; i < 5; i++ {
		writeTestFile(t, filepath.Join(dir, filepath.Base(t.Name())+string(rune('a'+i))+".md"), "content")
	}

	files, skipped, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  2,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 2 {
		t.Errorf("expected 2 files, got %d", len(files))
	}
	found := false
	for _, s := range skipped {
		if s == "max files limit (2) reached" {
			found = true
		}
	}
	if !found {
		t.Errorf("expected max files limit skip reason, got %v", skipped)
	}
}

func TestEnumerateFiles_HiddenFilesExcluded(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, ".hidden.md"), "hidden")
	writeTestFile(t, filepath.Join(dir, "visible.md"), "visible")

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file, got %d", len(files))
	}
}

func TestEnumerateFiles_GitignoreRespected(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, ".gitignore"), "node_modules\n*.log\n")
	writeTestFile(t, filepath.Join(dir, "node_modules", "pkg.txt"), "pkg")
	writeTestFile(t, filepath.Join(dir, "debug.log"), "logs")
	writeTestFile(t, filepath.Join(dir, "readme.md"), "readme")

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file (readme.md), got %d", len(files))
	}
}

func TestEnumerateFiles_RecursiveByDefault(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "top.md"), "top")
	writeTestFile(t, filepath.Join(dir, "sub", "nested.md"), "nested")

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file (non-recursive default), got %d", len(files))
	}
}

func TestEnumerateFiles_MaxFilesOver500Rejection(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "a.md"), "content")

	// EnumerateFiles respects maxFiles but does not enforce 500; the MCP tool
	// layer enforces the 500 cap. Verify that maxFiles=501 still works at the
	// library level (the restriction is in the tool, not here).
	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  501,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file, got %d", len(files))
	}
}

func TestEnumerateFiles_SymlinksSkipped(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "real.md"), "real file")

	// Create a symlink to a file outside the directory.
	outsideDir := t.TempDir()
	writeTestFile(t, filepath.Join(outsideDir, "secret.md"), "secret")
	symPath := filepath.Join(dir, "link.md")
	if err := os.Symlink(filepath.Join(outsideDir, "secret.md"), symPath); err != nil {
		t.Skipf("symlink not supported: %v", err)
	}

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	// Only the real file should be returned; symlink should be skipped.
	if len(files) != 1 {
		t.Errorf("expected 1 file (real.md only), got %d", len(files))
	}
	for _, f := range files {
		if filepath.Base(f.Path) == "link.md" {
			t.Error("symlink should have been skipped")
		}
	}
}

func TestEnumerateFiles_UnknownExtensionsSkipped(t *testing.T) {
	dir := t.TempDir()
	writeTestFile(t, filepath.Join(dir, "image.png"), "binary")
	writeTestFile(t, filepath.Join(dir, "doc.md"), "markdown")

	files, _, err := EnumerateFiles(context.Background(), EnumerateOptions{
		Directory: dir,
		Recursive: true,
		MaxFiles:  100,
	})
	if err != nil {
		t.Fatalf("enumerate: %v", err)
	}
	if len(files) != 1 {
		t.Errorf("expected 1 file, got %d", len(files))
	}
}
